diffkemp/simpll/PatternSet.cpp

```diff
@@ -0,0 +1,420 @@
+//===----------- PatternSet.cpp - Unordered set of code patterns ----------===//
+//
+//       SimpLL - Program simplifier for analysis of semantic difference      //
+//
+// This file is published under Apache 2.0 license. See LICENSE for details.
+// Author: Petr Silling, psilling@redhat.com
+//===----------------------------------------------------------------------===//
+///
+/// \file
+/// This file contains the implementation of the unordered LLVM code pattern
+/// set. Pattern sets are generated from the given pattern configuration file
+/// and hold all valid patterns that have been referenced there. Patterns may
+/// be instruction-based, or value-based. Instruction-based patterns are
+/// represented by multiple LLVM IR instructions, while value-based patterns
+/// contain only a single return instruction, which describes a difference in
+/// a single value.
+///
+//===----------------------------------------------------------------------===//
+
+#include "PatternSet.h"
+#include "Config.h"
+#include "ModuleAnalysis.h"
+#include <algorithm>
+#include <llvm/Support/Debug.h>
+#include <llvm/Support/Path.h>
+#include <llvm/Support/YAMLTraits.h>
+
+using namespace llvm::yaml;
+using namespace llvm::sys::path;
+
+// YAML to PatternConfiguration mapping
+namespace llvm {
+namespace yaml {
+template <> struct MappingTraits<PatternConfiguration> {
+    static void mapping(IO &IO, PatternConfiguration &Config) {
+        IO.mapOptional("on_parse_failure", Config.OnParseFailure);
+        IO.mapOptional("patterns", Config.PatternFiles);
+    }
+};
+} // namespace yaml
+} // namespace llvm
+
+/// Metadata operand offsets for different kinds of pattern metadata.
+const StringMap<int> PatternMetadata::MetadataOffsets = {
+        {"pattern-start", 1},
+        {"pattern-end", 1},
+        {"group-start", 1},
+        {"group-end", 1},
+        {"disable-name-comparison", 1},
+        {"not-an-input", 1}};
```

Comment:
Perhaps it'd be more clear to store the number of *operands* for each metadata type (and the offset is then *operands*+1).